### DISCUSSION TITLE
Efficiently create device with IP address via API

### DISCUSSION THREAD

**Discussion by llamafilm** [netbox-community/netbox#14643] [Help Wanted!] [answered]
I'm trying to build a script to help add batches of devices into Netbox when we commission a new room.  This process typically involves adding 10-15 new devices, each with a power port connection, ethernet switch connection, and primary IP.  I'm not seeing a way to create the device with a single API call; is this possible?  The [API schema docs](https://demo.netbox.dev/api/schema/swagger-ui/#/dcim/dcim_devices_create) show that `primary_ip4` is a valid key for a POST request to `/api/dcim/devices`, but when I try it fails with error: **"The specified IP address (192.168.5.1/24) is not assigned to this device."**

The interface is created automatically because it's part of the Device Type, so I was hoping this could all be handled in a single API call.  The alternative I found is convoluted and takes several API calls:
1. Create new device (POST /api/dcim/devices) and get its ID
2. Lookup interface filtered by name and device ID (GET /api/dcim/interfaces?device_id={x}&name={y})
3. Attach IP to Interface (PATCH /api/ipam/ip-addresses/{id})
4. Assign Primary IP (PATCH /api/dcim/devices/{id}

Then I would have to do a similar process to connect the power cable and ethernet cables.

**Comment by abhi1693** [comment, marked as answer]
You can only attach an IP address as primary if it exists on one of the interfaces for the device. So, your API calls look correct

**Comment by jburg379** [comment]
How did you get your script to automatically make new devices? I wrote a scrip to scan the network make the IPs on the netbox UI and its supposed to make new devices to go with the IPs but I keep getting this error. 
{"role":["This field is required."]}

**Reply by llamafilm** [reply]
The error is clearly stated, so I'm not sure what you're asking?  You need to define a role when creating a device.
